orchagent/nhgorch.cpp

```diff
@@ -0,0 +1,1199 @@
+#include "nhgorch.h"
+#include "switchorch.h"
+#include "neighorch.h"
+#include "crmorch.h"
+#include "routeorch.h"
+#include "bulker.h"
+#include "logger.h"
+#include "swssnet.h"
+
+extern sai_object_id_t gSwitchId;
+
+extern PortsOrch *gPortsOrch;
+extern CrmOrch *gCrmOrch;
+extern NeighOrch *gNeighOrch;
+extern SwitchOrch *gSwitchOrch;
+extern RouteOrch *gRouteOrch;
+
+extern size_t gMaxBulkSize;
+
+extern sai_next_hop_group_api_t* sai_next_hop_group_api;
+extern sai_next_hop_api_t*         sai_next_hop_api;
+extern sai_switch_api_t*            sai_switch_api;
+
+unsigned int NextHopGroup::m_count = 0;
+
+NhgOrch::NhgOrch(DBConnector *db, string tableName) :
+    Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+
+    /* Get the switch's maximum next hop group capacity. */
+    sai_attribute_t attr;
+    attr.id = SAI_SWITCH_ATTR_NUMBER_OF_ECMP_GROUPS;
+
+    sai_status_t status = sai_switch_api->get_switch_attribute(gSwitchId,
```

Comment:
Any reason for doing so? As I see it, routeOrch is the "manager" of ASIC route entry objects and should manage only such resources. nhgOrch is the "manager" of ASIC next hop group objects (so, ECMP groups). Having routeOrch retaining how many ECMP groups the ASIC supports would break the OOP principle of encapsulation, as routeOrch should not be responsible for knowing such things, but should rather query the appropriate nhgOrch about these things.

Reply:
Not really. As you explained in this comment, routeorch still owns the nexthop group, if `nhg_index `is empty. So we'vent totally isolated the ownership. In that aspect, I prefer to keep the old code retained. 

```
/*
 * Structure describing the next hop group used by a route.  As the next hop
 * groups can either be owned by RouteOrch or by NhgOrch, we have to keep track
 * of the next hop group index, as it is the one telling us which one owns it.
 */
struct RouteNhg
```

Reply:
I've moved this back to RouteOrch